Revamp close button component

mdo opened the pull request:
This PR redesigns our close buttons to make them more consistent across browsers and devices, as well as making it slightly easier to style them overall. With a little CSS `background-image` magic, we're sizing the &times; SVG to the `content-box`. This allows us to give use `padding` and rounded corners that carry the styling through with no impact to the &times; icon.

- Renamed `.close` to `.btn-close` for more specific class name
- Replaced inline `&times;` sign with embedded SVG in our CSS
- Improved focus styles with custom `box-shadow` "outline" like our buttons
- Increased focus shadow width of all buttons, including the new close button, from `.2em` to `.25em` for a more pixel perfect rendering
- Adds `.btn-close-white` variant for darker backgrounds.

Here's a preview of the default and disabled styles, as well as a focused and disabled style.

<img width="820" alt="Screen Shot 2020-09-13 at 7 12 22 PM" src="https://user-images.githubusercontent.com/98681/93036573-1ec15200-f5f5-11ea-980d-01acff0b6710.png">

<img width="820" alt="Screen Shot 2020-09-13 at 7 12 29 PM" src="https://user-images.githubusercontent.com/98681/93036572-1d902500-f5f5-11ea-99c9-0a2b84a74c79.png">

<img width="815" alt="Screen Shot 2020-09-13 at 7 47 56 PM" src="https://user-images.githubusercontent.com/98681/93038501-10296980-f5fa-11ea-99bc-eba5481479d9.png">

---

Closes #30927 and closes #31478.

Fixes #30917 (focus), fixes #28787 (default styling in toasts), fixes #27294 (rename). and fixes #27163 (focus).

Preview: <https://deploy-preview-31641--twbs-bootstrap.netlify.app/docs/5.0/components/close-button/>

mdo committed:
Rename .close to .btn-close, replace times sign with background SVG, update focus state

mdo committed:
Increase width of focus styles so it's more pixel perfect out of the box

mdo committed:
Increase margin in button examples to better highlight focus styles

mdo committed:
Add opacity variables, set opacity on disabled state

mdo committed:
Add details to the Migration page

mdo committed:
Tweak sizing and alignment of .btn-close in modal headers

mdo committed:
Update some close button examples

mdo committed:
Update toasts dismiss (moving more styles to CSS instead of utilities), fix some more close examples, and add a new .btn-close-white variant

mdo commented:
Without adding another variable, I think the only downside to the `invert()` on the white variant is the focus state gets inverted, too. I assumed this would only apply to the `background-image`. Might need to research more, or fallback to redeclaring the variable.

mdo committed:
Update .btn-close-white to use grayscale() filter too for better focus styles

XhmikosR commented:
@mdo: How does this look on Legacy Edge? As long as it sort of falls back to something and is acceptable, it's not a big deal if it's not supported.

https://caniuse.com/css-filters

mdo commented:
Looks good on Edge 18.

<img width="1552" alt="Screen Shot 2020-09-18 at 3 26 32 PM" src="https://user-images.githubusercontent.com/98681/93650455-619f7300-f9c3-11ea-858b-e1cdfad32180.png">

XhmikosR commented:
Maybe the `btn-close-white` focus border color could be a little whiter?

mdo committed:
Double the brightness for improved focus contrast

mdo commented:
> Maybe the `btn-close-white` focus border color could be a little whiter?

Added `brightness(200%)` to tackle that.

XhmikosR reviewed (approved):
LGTM, maybe wait for @twbs/css-review for another approval

ffoodd commented on scss/_alert.scss:
```diff
@@ -27,15 +27,14 @@
 
 .alert-dismissible {
-  padding-right: $close-font-size + $alert-padding-x * 2;
+  padding-right: $alert-padding-x * 3.5;
```
`3.5` and `1.25` above look  a bit random here. Does it match anything? Shouldn't they be handled as variables?

ffoodd commented on scss/_close.scss:
```diff
@@ -1,36 +1,41 @@
-  background-color: transparent;
-  border: 0;
+.btn-close-white {
+  filter: invert(1) grayscale(100%) brightness(200%);
```
Could it be useful to set this value as a variable, too?

ffoodd reviewed (approved):
Except a few hardcoded values, looks good to me 👌

mdo committed:
Add more variables, add a comment for the 3

mdo commented:
Nice, great comments @ffoodd! Updated the alert CSS to include two new variables. Also changed the `3.5` to `3` (previous value was too much) and explained why we use that value.
